### PR TITLE
Commitlog: Optionally support oversized entries using external file(s)

### LINKED ISSUE #18161: commitlog: store oversized mutations in blob files
Currently, commitlog throws an error if a mutation is larger than `max_mutation_size`.
In general, we do need to avoid large mutations, but we could support them more gracefully, if they are stored in a uniquely named blob file, which can be referred to in a corresponding commitlog record.

The blob file can be deleted once the commitlog segment that refers to it is recycled.

On startup, we should also detect and delete orphaned blob files that are not referred by any segment (maybe scylla crashed before committing the record referring to the blob, or lost it in some other way).

In addition, the blob sizes should be taken into account against the commitlog soft and hard limits.

### PR DESCRIPTION
**Pull request by elcallio**
Fixes #18161

Note: potentially temporary solution. This cannot handle hard size limit. 

Technically, if we could add some sort of sync between writers
we could fairly easily either write an oversized entry across several 
segments (if size is devided into more than one entry), or just allow a segment
to grow larger than normally allowed. However, such sync would probably impact 
throughput in the common case, and it would also be sensitive to 
both way to large temp buffer requirements (if we don't break each entry into 
chunks like we do here), but more importantly: If we crash while writing we
end up with half the entries written and any number missing -> partial 
recovery. 

We really want to know that the data is on disk fully before commiting
something that can be replayed to "real" commitlog. Hence we instead
write data to an "external" segment (just a special allocated segment), which 
we place no size limit on. We write each entry as a separate chunk to minimize
memory usage, then we flush and close the file. 

Once we are sure the data is on disk, we then just write a metadata entry
to the normal commitlog segment, in which we reference the external segment
and provide some other metadata (off, size). Then we set up a reference from
the "metedata" segment to the external one, so that once the metadata entries
are released, we also release the external segment.

Since we use a special prefix for the external file, much like recycled segments,
it will be picked up on replay, but ignored by the first level replay. If a metadata
entry is encountered however, we will find the external file and, if avail, replay the
data in it, pretending it comes from the "original" segment. I.e. transparent to 
client. The external files will then be freed/reused the same way any other replayed
segment is.

Again: this only works for non-hard limit CL.

### CODE REVIEW

File: db/commitlog/commitlog.cc [hunk 1]
```diff
@@ -3135,6 +3380,88 @@ db::commitlog::read_log_file(sstring filename, sstring pfx, commit_load_reader_f
...
+                                ++i;
+                            }
+                            // Now we have the metadata for the entry.
+                            if (i != j && std::get<1>(*i) == rp.pos) {
+                                if (std::get<2>(*i) != buf_rp.buffer.size_bytes()) {
+                                    throw std::runtime_error("Mismatched size in external entry");
+                                }
+                                // pretend the replay position is the reference location
+                                buf_rp.position = std::get<replay_position>(*i);
+                                co_await this->func(std::move(buf_rp));
```
Comment (denesb) [inline db/commitlog/commitlog.cc:3448]: Shouldn't this be just `func`?

File: db/commitlog/commitlog.cc [hunk 2]
```diff
@@ -1461,6 +1512,195 @@ future<R> db::commitlog::segment_manager::allocate_when_possible(T writer, db::t
...
 
+/**
+ * Write an oversized entry set as an external reference.
+ * 
+ * Note: potentially temporary solution. This cannot handle
+ * hard size limit. 
+ * 
+ * Technically, if we could add some sort of sync between writers
+ * we could fairly easily either write an oversized entry across several 
+ * segments (if size is devided into more than one entry), or just allow a segment
```
Comment (denesb) [inline db/commitlog/commitlog.cc:1523]: s/devided/divided/

File: db/commitlog/commitlog.cc [hunk 3]
```diff
@@ -1461,6 +1512,195 @@ future<R> db::commitlog::segment_manager::allocate_when_possible(T writer, db::t
...
+ * to grow larger than normally allowed. However, such sync would probably impact 
+ * throughput in the common case, and it would also be sensitive to 
+ * both way to large temp buffer requirements (if we don't break each entry into 
+ * chunks like we do here), but more importantly: If we crash while writing we
+ * end up with half the entries written and any number missing -> partial 
+ * recovery. 
+ * We really want to know that the data is on disk fully before commiting
```
Comment (denesb) [inline db/commitlog/commitlog.cc:1530]: s/commiting/committing/

**Review by denesb — COMMENTED** [review]
LGTM, but I am not that familiar with this code.
Out of curiosity, why won't the hard limit work with this? Is that even interesting for the schema commitlog (I think not)?

File: db/commitlog/commitlog.cc (hunk 1 above)
Comment (elcallio) [reply db/commitlog/commitlog.cc:3448]: local var has same name. Here we want to explicitly call the original (caller provided) callback.

**Comment by elcallio** [comment]
> Out of curiosity, why won't the hard limit work with this? Is that even interesting for the schema commitlog (I think not)?

It wont work because we can neither guarantee to fit an allocation into said limit, but more importantly, we have no way to within reasonable time ensure we can wait for "disk space" to become available in such as way that we can just write the new file.

File: db/commitlog/commitlog.cc (hunk 1 above)
Comment (elcallio) [reply db/commitlog/commitlog.cc:3448]: But you are right, here we can just do "func".

**Comment by michoecho** [comment]
Doesn't this change need some additional treatment (i.e. a cluster feature flag) to support Scylla version rollback without dropping committed mutations?

**Comment by michoecho** [comment]
> In general, we can not do neither upgrade or rollback with a dirty commit log. If you want to roll back, do a clean startup + shutdown first.

That's surprising. Is it enforced somewhere, or at least documented? Also, does this mean that in the case when we want to roll back a new version that crashes on startup (so it can't perform a clean shutdown), we are doomed?

**Comment by avikivity** [comment]
> Note: potentially temporary solution. This cannot handle hard size limit.


We can't add temporary solutions to on-disk structures. They're persistent.

Why not allow spreading commitlog entries over multiple segments?

**Comment by elcallio** [comment]
> > 
> 
> We can't add temporary solutions to on-disk structures. They're persistent.

I think one of the more defining characteristics of the commit log is that it is at best transient. I.e. never persists across an actual boot cycle. So I would argue that a solution that might at some point be superseded is acceptable.

> 
> Why not allow spreading commitlog entries over multiple segments?

Because if one does, then we might write part of the entry list into what is/must be essentially valid data entries, fail at point X, and report failure to caller, while still mutations remain in the log. But more importantly, we'd have to do some sort of locking across continuation points, because we'd need to acquire N segments (for M bytes placement), while competing with other incoming writes, etc. 

This assumes the oversized write is just the sum of a multi-entry write mind you. In which case, absolute ordering _could_ be ignored, and the only real issue is whether or not we wrote parts of the entries before failing/crashing. 
If it is in fact a single, huge mutation, we need to ensure X consecutive segments reserved across the full write, essentially locking out all other writes, and then we need to keep track of an internal part list somehow (complicated since we probably have not written all data while writing the header). 

This PR:s solution has the benefit of us being able to fully write the large entry list, ensure it being flushed proper, then write the meta entry referencing the actual data, and then either succeed or fail, in which case the data can be kept or discarded as is proper.

**Comment by elcallio** [comment]
> > 
> 
> That's surprising. Is it enforced somewhere, or at least documented? Also, does this mean that in the case when we have to roll back a new version that crashes on startup (so it can't perform a clean shutdown), we are doomed?

If a new version crashes on startup, it seems unlikely that there would be data in the commitlog to play back.
Cluster feature flag is pointless, since the files are 100% node local. 
Regardless, if you have commitlog data and try to start with a rollbacked version that cannot read whatever format change was made, you will fail. Beyond never making changes to the file format, I see no real way of handling that beyond enforcing clean shutdowns when up/downgrading.

**Comment by michoecho** [comment]
> If a new version crashes on startup, it seems unlikely that there would be data in the commitlog to play back.

@elcallio "Seems unlikely" isn't good enough. Is it really that unlikely that a new Scylla version could keep crashing at a point after writing some mutation to commitlog, but before the administrator has a chance to shut it down cleanly?

> Cluster feature flag is pointless, since the files are 100% node local.

No, because cluster features are exactly the (documented) points of no return for rollbacks. I.e: by design, we allow a rollback if and only if a new cluster feature hasn't been enabled yet — when the old version sees that an unknown (future) feature flag has been previously enabled on the node, the node will refuse to boot, before it even attempts to replay the commitlog.

So if you only start writing the new commitlog format after a corresponding feature flag is enabled, you will guarantee that it won't be read by a Scylla version which doesn't understand it.

> Regardless, if you have commitlog data and try to start with a rollbacked version that cannot read whatever format change was made, you will fail. Beyond never making changes to the file format, I see no real way of handling that beyond enforcing clean shutdowns when up/downgrading.

Obviously you can't support any rollback scenario possible. But the point of no return should be documented, and — most importantly — enforced by the software, not by the administrator.

> If you want to roll back, do a clean startup + shutdown first.

If this isn't automatically enforced anywhere, then you are possibly the only person in the world who knows that rolling back a Scylla node after an unclean shutdown is forbidden, and allowed to corrupt the cluster.

**Comment by mykaul** [comment]
> If this isn't automatically enforced anywhere, then you are possibly the only person in the world who knows that rolling back a Scylla node after an unclean shutdown is forbidden, and allowed to corrupt the cluster.

Our support and cloud team also know how to do it (and so do I, as I remember a previous conversation about it - when discussing v2 -> v3 format discussion), so he's not the only person in the world.

**Comment by elcallio** [comment]
> > If a new version crashes on startup, it seems unlikely that there would be data in the commitlog to play back.
> 
> @elcallio "Seems unlikely" isn't good enough. Is it really that unlikely that a new Scylla version could keep crashing at a point after writing some mutation to commitlog, but before the administrator has a chance to shut it down cleanly?
> 
Yes. More to the point: If the server could get to a point where it wrote to the commit log before crashing, then it can reach a point where said CL is replayed, but we don't yet accept client writes. (Schema modifications are a special case here, but if you crash with only a schema change not applied, but attempted, rolling back w/o said CL data should be fine). 

> > Cluster feature flag is pointless, since the files are 100% node local.
> 
> No, because cluster features are exactly the (documented) points of no return for rollbacks. I.e: by design, we allow a rollback if and only if a new cluster feature hasn't been enabled yet — when the old version sees that an unknown (future) feature flag has been previously enabled on the node, the node will refuse to boot, before it even attempts to replay the commitlog.
> 
So now the whole cluster crashes on startup? I fail to see an improvement. 

> So if you only start writing the new commitlog format after a corresponding feature flag is enabled, you will guarantee that it won't be read by a Scylla version which doesn't understand it.
> 
Then if you really want, you can place the enabling of the feaure (in schema commitlog creation) behind a flag, but again, a schema change in a new version should probably not be applicable to a rolled back version, so again, not much point. 

> > If you want to roll back, do a clean startup + shutdown first.
> 
> If this isn't automatically enforced anywhere, then you are possibly the only person in the world who knows that rolling back a Scylla node after an unclean shutdown is forbidden, and allowed to corrupt the cluster.

"automatically enforced"? Also, just because you don't know something does not mean no other person does. I would hope that most of our customer facing engs. are aware that upgrade == clean node.

**Comment by mykaul** [comment]
@elcallio - did you look at dtest failures? Are you re-running them?

**Comment by elcallio** [comment]
No. This PR is (until otherwise countermanded) probably superseded by #19472

**Comment by elcallio** [comment]
Closing in favour of #19472

**Pull request closed without merging**